sdk/core/azure-core/src/main/java/com/azure/core/util/polling/SimpleSyncPoller.java

```diff
@@ -0,0 +1,284 @@
+            this.terminalPollContext = this.pollingContext.copy();
+        }
+        return response;
+    }
+
+    @Override
+    public PollResponse<T> waitForCompletion() {
+        PollingContext<T> currentTerminalPollContext = this.terminalPollContext;
+        if (currentTerminalPollContext != null) {
+            return currentTerminalPollContext.getLatestResponse();
+        } else {
+            PollingContext<T> context = this.pollingContext.copy();
+            SyncPollResponse<T, U> finalAsyncPollResponse = pollingLoopSync(context);
+            PollResponse<T> response = toPollResponse(finalAsyncPollResponse);
+            this.terminalPollContext = context;
+            return response;
+        }
+    }
+
+    @Override
+    public PollResponse<T> waitForCompletion(Duration timeout) {
+        PollingContext<T> currentTerminalPollContext = this.terminalPollContext;
+        if (currentTerminalPollContext != null) {
+            return currentTerminalPollContext.getLatestResponse();
+        } else {
+            PollingContext<T> context = this.pollingContext.copy();
+            SyncPollResponse<T, U> finalSyncPollResponse = pollingLoopSync(context, timeout);
+            PollResponse<T> response = toPollResponse(finalSyncPollResponse);
+            this.terminalPollContext = context;
+            return response;
+        }
+    }
+
+    @Override
+    public PollResponse<T> waitUntil(LongRunningOperationStatus statusToWaitFor) {
+        Objects.requireNonNull(statusToWaitFor, "'statusToWaitFor' cannot be null.");
+        PollingContext<T> currentTerminalPollContext = this.terminalPollContext;
+        if (currentTerminalPollContext != null
+                && currentTerminalPollContext.getLatestResponse().getStatus() == statusToWaitFor) {
+            return currentTerminalPollContext.getLatestResponse();
+        } else {
+            PollingContext<T> context = this.pollingContext.copy();
+            SyncPollResponse<T, U> syncPollResponse = pollingLoopSync(context, statusToWaitFor);
+
+            if (!syncPollResponse.getStatus().equals(statusToWaitFor)) {
+                throw LOGGER.logExceptionAsError(new NoSuchElementException("Polling completed without"
+                    + " receiving the given status '" + statusToWaitFor + "'."));
+            }
+            PollResponse<T> response = toPollResponse(syncPollResponse);
+            if (response.getStatus().isComplete()) {
+                this.terminalPollContext = context;
+            }
+            return response;
+        }
+    }
+
+    @Override
+    public PollResponse<T> waitUntil(Duration timeout, LongRunningOperationStatus statusToWaitFor) {
+        Objects.requireNonNull(timeout, "'timeout' cannot be null.");
+        if (timeout.isNegative() || timeout.isZero()) {
+            throw LOGGER.logExceptionAsWarning(new IllegalArgumentException(
+                "Negative or zero value for timeout is not allowed."));
+        }
```

Comment:
We should have this check for `waitForCompletion(Duration timeout)` overload too if `timeout` is not null.